File: packages/js-sdk/examples/turboquote-pricebooks.ts
/**
 * TurboQuote Example: Price Books & Quote Pricing
 *
 * Fully self-contained — creates all data it needs, then cleans up.
 * Just add your API key and run.
 *
 * Methods demonstrated:
 * - configure()
 * - createType(), deleteType()
 * - createProduct(), deleteProduct()
 * - createCompany(), deleteCompany()
 * - createContact(), deleteContact()
 * - createPriceBook(), listPriceBooks(), getPriceBook(), updatePriceBook()
 * - listPriceBookProducts(), duplicatePriceBook(), deletePriceBook()
 * - createQuote(), applyPriceBook(), removePriceBook(), deleteQuote()
 * - sendQuoteWithDeliverable() (optional — requires a TurboDocx deliverable)
 *
 * Run: npx tsx examples/turboquote-pricebooks.ts
 */

import { TurboQuote } from '@turbodocx/sdk';

async function priceBookExample(): Promise<void> {
  // =============================================
  // 1. CONFIGURE
  // =============================================
  TurboQuote.configure({
    apiKey: process.env.TURBODOCX_API_KEY || 'your-api-key-here',
    orgId: process.env.TURBODOCX_ORG_ID || 'your-org-id-here',
    baseUrl: process.env.TURBODOCX_BASE_URL || 'https://api.turbodocx.com',
  });

  try {
    // =============================================
    // 2. SET UP — categories, products, CRM
    // =============================================
    console.log('2. Setting up prerequisites...');

    const pbType = await TurboQuote.createType({
      name: 'Customer Tier',
      categoryType: 'pricebook_type',
    });
    console.log(`  Price book type: ${pbType.name}`);

    const productCategory = await TurboQuote.createType({
      name: 'Cloud Services',
      categoryType: 'product_category',
    });

    const product1 = await TurboQuote.createProduct({
      name: 'Cloud Storage (1TB)',
      listPrice: 99.99,
      billingFrequency: 'monthly',
      sku: 'CLD-STR-1TB',
      categoryId: productCategory.id,
      currency: 'USD',
    });
    console.log(`  Product 1: ${product1.name} — $${product1.listPrice}/mo`);

    const product2 = await TurboQuote.createProduct({
      name: 'API Gateway',
      listPrice: 199.99,
      billingFrequency: 'monthly',
      sku: 'API-GW-001',
      categoryId: productCategory.id,
      currency: 'USD',
    });
    console.log(`  Product 2: ${product2.name} — $${product2.listPrice}/mo`);

    const company = await TurboQuote.createCompany({
      name: 'PriceBook Demo Corp',
      contacts: [{ name: 'Sam Buyer', email: 'sam@demo.com' }],
    });

    const contact = await TurboQuote.createContact({
      name: 'Sam Buyer',
      companyId: company.id,
      email: 'sam@demo.com',
    });
    console.log(`  Company: ${company.name}, Contact: ${contact.name}\n`);

    // =============================================
    // 3. CREATE A PRICE BOOK
    // =============================================
    console.log('3. Creating price book...');

    const priceBook = await TurboQuote.createPriceBook({
      name: 'Enterprise Tier',
      priceBookTypeId: pbType.id,
      description: 'Discounted pricing for enterprise customers',
      discountPercent: 15,
      validFrom: '2026-01-01',
      validTo: '2026-12-31',
      showInQuoteBuilder: true,
      productPricing: [
        { productId: product1.id, discountPercent: 20 },
        { productId: product2.id, finalPrice: 149.99 },
      ],
    });

    console.log(`  ${priceBook.name}: ${priceBook.discountPercent}% default discount`);
    console.log(`  Valid: ${priceBook.validFrom} to ${priceBook.validTo}\n`);

    // =============================================
    // 4. LIST & GET PRICE BOOKS
    // =============================================
    console.log('4. Browsing price books...');

    const priceBooks = await TurboQuote.listPriceBooks({ showInQuoteBuilder: true, limit: 10 });
    console.log(`  ${priceBooks.totalRecords} price book(s) visible in quote builder`);

    const pbDetail = await TurboQuote.getPriceBook(priceBook.id);
    console.log(`  ${pbDetail.name}: ${pbDetail.productPricing?.length || 0} product overrides\n`);

    // =============================================
    // 5. UPDATE PRICE BOOK
    // =============================================
    console.log('5. Updating price book...');

    const updatedPB = await TurboQuote.updatePriceBook(priceBook.id, {
      discountPercent: 20,
      description: 'Enterprise tier — increased discount for H2 2026',
      productPricing: [
        { productId: product1.id, discountPercent: 25 },
        { productId: product2.id, finalPrice: 139.99 },
      ],
    });
    console.log(`  New default discount: ${updatedPB.discountPercent}%\n`);

    // =============================================
    // 6. LIST PRICE BOOK PRODUCTS
    // =============================================
    console.log('6. Listing price book products...');

    const pbProducts = await TurboQuote.listPriceBookProducts(priceBook.id, { limit: 10 });
    console.log(`  ${pbProducts.totalRecords} product pricing override(s)`);
    for (const pp of pbProducts.results) {
      console.log(`    - Product ${pp.productId}: ${pp.discountPercent}% off → $${pp.finalPrice}`);
    }
    console.log();

    // =============================================
    // 7. DUPLICATE PRICE BOOK
    // =============================================
    console.log('7. Duplicating price book...');

    const seasonalPB = await TurboQuote.duplicatePriceBook(priceBook.id);
    console.log(`  Copy: ${seasonalPB.name} (${seasonalPB.id})\n`);

    // =============================================
    // 8. APPLY PRICE BOOK TO QUOTE
    // =============================================
    console.log('8. Applying price book to a quote...');

    const quote = await TurboQuote.createQuote({
      name: 'Enterprise Quote — Demo Corp',
      companyId: company.id,
      contactId: contact.id,
      currency: 'USD',
      termDays: 30,
    });
    console.log(`  Quote: ${quote.quoteNumber}`);

    const quotePB = await TurboQuote.applyPriceBook(quote.id, priceBook.id);
    console.log(`  Applied price book: ${quotePB.priceBookId}\n`);

    // =============================================
    // 9. SEND QUOTE WITH DELIVERABLE (optional)
    // =============================================
    // sendQuoteWithDeliverable merges a TurboDocx-generated document
    // (e.g., a cover letter or SOW) with the quote PDF before sending.
    // Requires a real deliverableId — skip gracefully if not available.
    if (process.env.DELIVERABLE_ID) {
      console.log('9. Sending quote with deliverable...');
      const sendResult = await TurboQuote.sendQuoteWithDeliverable(quote.id, {
        deliverableId: process.env.DELIVERABLE_ID,
        mergePosition: 'beginning',
      });
      console.log(`  ✅ ${sendResult.message}`);
      if (sendResult.signatureDocumentId) {
        console.log(`  Signature Doc: ${sendResult.signatureDocumentId}`);
      }
      console.log();
    } else {
      console.log('9. Skipping sendQuoteWithDeliverable (set DELIVERABLE_ID env var to test)\n');
    }

    // =============================================
    // 10. REMOVE PRICE BOOK FROM QUOTE
    // =============================================
    console.log('10. Removing price book from quote...');

    const quoteNoPB = await TurboQuote.removePriceBook(quote.id);
    console.log(`  priceBookId is now: ${quoteNoPB.priceBookId}\n`);

    // =============================================
    // 11. CLEANUP
    // =============================================
    console.log('11. Cleaning up...');

    await TurboQuote.deleteQuote(quote.id);
    await TurboQuote.deletePriceBook(seasonalPB.id);
    await TurboQuote.deletePriceBook(priceBook.id);
    await TurboQuote.deleteContact(contact.id);
    await TurboQuote.deleteCompany(company.id);
    await TurboQuote.deleteProduct(product2.id);
    await TurboQuote.deleteProduct(product1.id);
    await TurboQuote.deleteType(productCategory.id);
    await TurboQuote.deleteType(pbType.id);

    console.log('  ✅ All test data removed');
    console.log('\n=== Price book example completed successfully! ===');
  } catch (error: any) {
    console.error(`Error: ${error.message}`);
    if (error.statusCode) {
      console.error(`Status Code: ${error.statusCode}`);
    }
  }
}

priceBookExample();

File: packages/go-sdk/examples/turboquote_pricebooks.go
//go:build ignore
// +build ignore

// TurboQuote Price Books Example
//
// Demonstrates price book management:
//   - Create a price book type and product category (via CreateType)
//   - Create a product
//   - Create a price book with product-level pricing
//   - Update the price book (PATCH)
//   - Create a quote, add a line item, and apply the price book
//   - Optionally send quote with deliverable (set DELIVERABLE_ID env var)
//   - Clean up
//
// Usage:
//
//	TURBODOCX_API_KEY=TDX-... TURBODOCX_ORG_ID=org_... go run turboquote_pricebooks.go
//	DELIVERABLE_ID=del_... go run turboquote_pricebooks.go   # also test SendQuoteWithDeliverable

package main

import (
	"context"
	"fmt"
	"os"

	turbodocx "github.com/TurboDocx/SDK/packages/go-sdk"
)

func main() {
	ctx := context.Background()

	client, err := turbodocx.NewQuoteClient(turbodocx.QuoteClientConfig{
		APIKey: getEnv("TURBODOCX_API_KEY", ""),
		OrgID:  getEnv("TURBODOCX_ORG_ID", ""),
	})
	if err != nil {
		fmt.Printf("Error creating client: %v\n", err)
		os.Exit(1)
	}

	// --- 1. Create a price book type ---
	fmt.Println("Creating price book type...")
	pbType, err := client.CreateType(ctx, &turbodocx.CreateQuoteTypeRequest{
		Name:         "Enterprise Tiers",
		CategoryType: turbodocx.CategoryTypePricebookType,
	})
	if err != nil {
		fmt.Printf("Error creating price book type: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Created price book type: %s (%s)\n", pbType.Name, pbType.ID)
	defer func() {
		fmt.Println("\nCleaning up price book type...")
		if _, err := client.DeleteType(ctx, pbType.ID); err != nil {
			fmt.Printf("Warning: failed to delete price book type: %v\n", err)
		}
	}()

	// --- 2. Create a product category and product ---
	fmt.Println("Creating product category...")
	productCat, err := client.CreateType(ctx, &turbodocx.CreateQuoteTypeRequest{
		Name:         "Cloud Services",
		CategoryType: turbodocx.CategoryTypeProductCategory,
	})
	if err != nil {
		fmt.Printf("Error creating product category: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Created product category: %s (%s)\n", productCat.Name, productCat.ID)
	defer func() {
		fmt.Println("Cleaning up product category...")
		if _, err := client.DeleteType(ctx, productCat.ID); err != nil {
			fmt.Printf("Warning: failed to delete product category: %v\n", err)
		}
	}()

	fmt.Println("Creating product...")
	product, err := client.CreateProduct(ctx, &turbodocx.CreateProductRequest{
		Name:             "Cloud Storage 1TB",
		ListPrice:        120.00,
		BillingFrequency: "annual",
		CategoryID:       productCat.ID,
	})
	if err != nil {
		fmt.Printf("Error creating product: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Created product: %s @ $%.2f/year (%s)\n", product.Name, product.ListPrice, product.ID)
	defer func() {
		fmt.Println("Cleaning up product...")
		if _, err := client.DeleteProduct(ctx, product.ID); err != nil {
			fmt.Printf("Warning: failed to delete product: %v\n", err)
		}
	}()

	// --- 3. Create a price book with product-level pricing ---
	fmt.Println("Creating price book...")
	discountType := turbodocx.DiscountTypePercent
	discountPct := 15.0
	pricebook, err := client.CreatePriceBook(ctx, &turbodocx.CreatePriceBookRequest{
		Name:            "Enterprise 15% Off",
		PriceBookTypeID: pbType.ID,
		ValidFrom:       "2025-01-01",
		DiscountPercent: &discountPct,
		ProductPricing: []turbodocx.PriceBookProductPricingInput{
			{
				ProductID:       product.ID,
				DiscountType:    &discountType,
				DiscountPercent: &discountPct,
			},
		},
	})
	if err != nil {
		fmt.Printf("Error creating price book: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Created price book: %s (%s)\n", pricebook.Name, pricebook.ID)
	defer func() {
		fmt.Println("Cleaning up price book...")
		if _, err := client.DeletePriceBook(ctx, pricebook.ID); err != nil {
			fmt.Printf("Warning: failed to delete price book: %v\n", err)
		}
	}()

	// --- 4. Update the price book (demonstrate PATCH with nullable field) ---
	fmt.Println("Updating price book description...")
	desc := "Preferred pricing for enterprise accounts"
	updatedPB, err := client.UpdatePriceBook(ctx, pricebook.ID, &turbodocx.UpdatePriceBookRequest{
		Description: &desc,
	})
	if err != nil {
		fmt.Printf("Warning: update price book error: %v\n", err)
	} else {
		fmt.Printf("Updated price book description: %v\n", ptrStr(updatedPB.Description))
	}

	// --- 5. Create a company, contact, and quote ---
	company, err := client.CreateCompany(ctx, &turbodocx.CreateCompanyRequest{Name: "Enterprise Corp"})
	if err != nil {
		fmt.Printf("Error creating company: %v\n", err)
		os.Exit(1)
	}
	defer func() {
		client.DeleteCompany(ctx, company.ID) //nolint:errcheck
	}()

	contactEmail := "enterprise@corp.example.com"
	contact, err := client.CreateContact(ctx, &turbodocx.CreateContactRequest{
		Name:      "Enterprise Contact",
		CompanyID: company.ID,
		Email:     &contactEmail,
	})
	if err != nil {
		fmt.Printf("Error creating contact: %v\n", err)
		os.Exit(1)
	}
	defer func() {
		client.DeleteContact(ctx, contact.ID) //nolint:errcheck
	}()

	fmt.Println("Creating quote...")
	currency := "USD"
	quote, err := client.CreateQuote(ctx, &turbodocx.CreateQuoteRequest{
		Name:         "Enterprise Quote",
		CompanyID:    company.ID,
		ContactID:    contact.ID,
		CurrencyCode: &currency,
	})
	if err != nil {
		fmt.Printf("Error creating quote: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Created quote: %s (%s)\n", quote.Name, quote.ID)
	defer func() {
		fmt.Println("Cleaning up quote...")
		if _, err := client.DeleteQuote(ctx, quote.ID); err != nil {
			fmt.Printf("Warning: failed to delete quote: %v\n", err)
		}
	}()

	// Add the product as a line item
	lineItems, err := client.AddLineItems(ctx, quote.ID, turbodocx.AddLineItemRequest{
		ProductID:        &product.ID,
		ProductName:      product.Name,
		UnitPrice:        product.ListPrice,
		BillingFrequency: "annual",
	})
	if err != nil {
		fmt.Printf("Error adding line item: %v\n", err)
		os.Exit(1)
	}
	if len(lineItems) > 0 {
		fmt.Printf("Added line item: %s (subtotal $%.2f)\n",
			ptrStr(lineItems[0].ProductName), lineItems[0].Subtotal)
	}

	// --- 6. Apply price book ---
	fmt.Println("Applying price book to quote...")
	applyResp, err := client.ApplyPriceBook(ctx, quote.ID, pricebook.ID)
	if err != nil {
		fmt.Printf("Warning: apply price book error: %v\n", err)
	} else {
		fmt.Printf("Price book applied! Updated: %d, Skipped: %d\n",
			applyResp.UpdatedCount, applyResp.SkippedCount)
		fmt.Printf("Quote grand total after discount: $%.2f\n", applyResp.QuoteResult.GrandTotal)
	}

	// --- 7. Optional: Send with deliverable ---
	if deliverableID := os.Getenv("DELIVERABLE_ID"); deliverableID != "" {
		fmt.Printf("\nSending quote with deliverable %s...\n", deliverableID)
		sendResp, err := client.SendQuoteWithDeliverable(ctx, quote.ID, &turbodocx.SendQuoteWithDeliverableRequest{
			DeliverableID: deliverableID,
			MergePosition: "append",
		})
		if err != nil {
			fmt.Printf("Error sending: %v\n", err)
		} else {
			fmt.Printf("Sent! Status: %s, Document ID: %s\n",
				sendResp.QuoteResult.Status, sendResp.DocumentID)
		}
	}

	fmt.Println("\nDone!")
}

func getEnv(key, fallback string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return fallback
}

func ptrStr(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}
